rivertype/river_type.go

```diff
@@ -229,6 +230,47 @@ type AttemptError struct {
 	Trace string `json:"trace"`
 }
 
+type JobInsertParams struct {
+	Args         JobArgs
+	CreatedAt    *time.Time
+	EncodedArgs  []byte
+	Kind         string
+	MaxAttempts  int
+	Metadata     []byte
+	Priority     int
+	Queue        string
+	ScheduledAt  *time.Time
+	State        JobState
+	Tags         []string
+	UniqueKey    []byte
+	UniqueStates byte
+}
+
+// JobMiddleware provides an interface for middleware that integrations can use
+// to encapsulate common logic around various phases of a job's lifecycle.
+//
+// Implementations should embed river.JobMiddlewareDefaults to inherit default
+// implementations for phases where no custom code is needed, and for forward
+// compatibility in case new functions are added to this interface.
+type JobMiddleware interface {
+	// InsertMany is invoked around a batch insert operation. Implementations
+	// must always include a call to doInner to call down the middleware stack
+	// and perfom the batch insertion, and may run custom code before and after.
+	//
+	// Returning an error from this function will fail the overarching insert
+	// operation, even if the inner insertion originally succeeded.
+	InsertMany(ctx context.Context, manyParams []*JobInsertParams, doInner func(ctx context.Context) ([]*JobInsertResult, error)) ([]*JobInsertResult, error)
```

Comment:
I think it could still be possible to have a single-job method here for simplicity, but we would need to have some adapter code to make that work. It seemed easier to get this out with a more minimal version that does just what we need for now. Plus `JobMiddlewareDefaults` helps enough with forward compatibility if we update the interface.